drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+	if (ret) {
+		fprintf(stderr, "cannot create framebuffer for connector %u\n",
+			conn->connector_id);
+		modeset_destroy_fb(fd, &dev->bufs[0]);
+		return ret;
+	}
+
+	return 0;
+}
+
+/*
+ * modeset_find_crtc() stays the same.
+ */
+
+static int modeset_find_crtc(int fd, drmModeRes *res, drmModeConnector *conn,
+			     struct modeset_dev *dev)
+{
+	drmModeEncoder *enc;
+	unsigned int i, j;
+	uint32_t crtc;
+	struct modeset_dev *iter;
+
+	/* first try the currently conected encoder+crtc */
+	if (conn->encoder_id)
+		enc = drmModeGetEncoder(fd, conn->encoder_id);
+	else
+		enc = NULL;
+
+	if (enc) {
+		if (enc->crtc_id) {
+			crtc = enc->crtc_id;
+			for (iter = modeset_list; iter; iter = iter->next) {
+				if (iter->crtc_id == crtc) {
+					crtc = -1;
+					break;
+				}
+			}
+
+			if (crtc > 0) {
+				drmModeFreeEncoder(enc);
+				dev->crtc_id = crtc;
+				return 0;
+			}
+		}
+
+		drmModeFreeEncoder(enc);
+	}
+
+	/* If the connector is not currently bound to an encoder or if the
+	 * encoder+crtc is already used by another connector (actually unlikely
+	 * but lets be safe), iterate all other available encoders to find a
+	 * matching CRTC.
+	 */
+	for (i = 0; i < conn->count_encoders; ++i) {
+		enc = drmModeGetEncoder(fd, conn->encoders[i]);
+		if (!enc) {
+			fprintf(stderr, "cannot retrieve encoder %u:%u (%d): %m\n",
+				i, conn->encoders[i], errno);
+			continue;
+		}
+
+		/* iterate all global CRTCs */
+		for (j = 0; j < res->count_crtcs; ++j) {
+			/* check whether this CRTC works with the encoder */
+			if (!(enc->possible_crtcs & (1 << j)))
+				continue;
+
+			/* check that no other device already uses this CRTC */
+			crtc = res->crtcs[j];
+			for (iter = modeset_list; iter; iter = iter->next) {
+				if (iter->crtc_id == crtc) {
+					crtc = -1;
+					break;
+				}
+			}
+
+			/* We have found a CRTC, so save it and return. Note that we have to
+			 * save its index as well. The CRTC index (not its ID) will be used
+			 * when searching for a suitable plane. TODO: the index is not
+			 * really persistent?
+			 */
+			if (crtc >= 0) {
+				fprintf(stdout, "crtc %u found for encoder %u, will need full modeset\n",
+					crtc, conn->encoders[i]);;
+				drmModeFreeEncoder(enc);
+				dev->crtc_id = crtc;
+				dev->crtc_index = j;
+				return 0;
+			}
+		}
+
+		drmModeFreeEncoder(enc);
+	}
+
+	fprintf(stderr, "cannot find suitable crtc for connector %u\n",
+		conn->connector_id);
+	return -ENOENT;
+}
+
+/*
+ * get_property_value() is a new function. Given a driver, the properties of
+ * an object and a name, search for the value of property 'name'.
+ */
+
+static uint64_t get_property_value(int fd, drmModeObjectPropertiesPtr props, 
+					const char *name)
+{
+	drmModePropertyPtr prop;
+	uint64_t value;
+	int j;
+	
+	for (j = 0; j < props->count_props; j++) {
+		prop = drmModeGetProperty(fd, props->props[j]);
+		if (!strcmp(prop->name, name))
+			value = props->prop_values[j];
+		drmModeFreeProperty(prop);
+	}
+	return value;
```

Comment:
If the name never matches, this returns an uninitialized value.

Reply:
Good catch! Fixed.